### PR TITLE
Budget alerts

### PR DESCRIPTION
**Pull request by koopmant**
related to https://github.com/DIAGNijmegen/rse-grand-challenge-admin/issues/319

### CODE REVIEW

File: app/grandchallenge/challenges/emails.py [hunk 1]
```diff
@@ -91,3 +91,24 @@ def send_challenge_status_update_email(challengerequest, challenge=None):
...
+            challenge_name=challenge.short_name,
+            warning_threshold=warning_threshold,
+        ),
+        markdown_message=format_html(
+            "We would like to inform you that {percent_budget_consumed}% of the "
+            "compute budget for your challenge has been used.",
+            percent_budget_consumed=challenge.percent_budget_consumed,
+        ),
+        recipients=[
+            *challenge.get_admins(),
```
Comment (koopmant) [inline app/grandchallenge/challenges/emails.py:110]: Initially I had the `challenge.creator` here explicitly. But I think they are always in the admins group? If I include the creator here separately I end up sending the email to them twice.

Comment (jmsmkn) [reply app/grandchallenge/challenges/emails.py:110]: The creator is just for tracking so should be treated as anything special. Only the admins group should be emailed, think of the following situations:

- We create a challenge on behalf of another user, we are then removed from the admins group.
- Someone creates a challenge, later leaves the organisation and is removed from the admins group by the other admins.

In both situations we do not want the creator to get an email. They will get an email if they are part of the admins group, which is what we want.

File: app/grandchallenge/challenges/models.py [hunk 2]
```diff
@@ -707,7 +725,7 @@ def should_be_open_but_is_over_budget(self):
             for phase in self.phase_set.all()
         )
 
-    @cached_property
+    @property
```
[the remark this answers is not in the document]
Comment (jmsmkn) [reply app/grandchallenge/challenges/models.py:728]: I think it might be good to keep this as `approved_compute_costs_euro_millicents` involves a database query. To ensure the value you can send it to the task as an argument. But, we probably do not need such a level of detail in the email, just sending the threshold crossed would be okay.

File: app/grandchallenge/challenges/emails.py (hunk 1 above)
Comment (amickan) [reply app/grandchallenge/challenges/emails.py:110]: Sending this to the admins group is fine, I think. The creator is indeed automatically added to the admins and will stay an admin unless another admin removes him/her, in which case you wouldn't want to email them.

File: app/grandchallenge/challenges/models.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (amickan) [reply app/grandchallenge/challenges/models.py:728]: I think this is used in two templates, the challenge statistics page and the challenge cost overview. I'm not 100% sure but I think making this a normal property would mean recalculating this every time we search or filter or order the cost overview table, for example, which would not be desirable.

File: app/grandchallenge/challenges/tasks.py [hunk 3]
```diff
@@ -55,12 +58,34 @@ def update_challenge_results_cache():
...
+        for threshold in sorted(
+            challenge.percent_budget_consumed_warning_thresholds, reverse=True
+        ):
+            if (
+                challenge.initial_value("compute_cost_euro_millicents")
+                < challenge.approved_compute_costs_euro_millicents
+                * threshold
+                / 100
+                <= challenge.compute_cost_euro_millicents
+            ):
```
Comment (jmsmkn) [inline app/grandchallenge/challenges/tasks.py:75]: I find this if statement hard to grok. Can you maybe assign some variables to help with readability? Style wise, it is usually easiest for other developers to have everything laid out for them, and the if statement be quite straightforward:

```python
if should_do_something:
    do_something()
```

then `should_do_something` is built up through logic:

```python
it_is_sunday = datetime.datetime.now().weekday() == 6
feeling_under_the_weather = random.random() < 0.05

should_do_something = not it_is_sunday and not feeling_under_the_weather
```

File: app/tests/challenges_tests/test_tasks.py [hunk 4]
```diff
@@ -121,3 +131,42 @@ def test_challenge_request_budget_calculation(settings):
...
+        storage_costs_euros=0,
+        payment_status=PaymentStatusChoices.PAID,
+    )
+    phase = PhaseFactory(challenge=challenge)
+    EvaluationFactory(
+        submission__phase=phase,
+        compute_cost_euro_millicents=800000,
+        time_limit=60,
+    )
+    update_compute_costs_and_storage_size()
```
[the remark this answers is not in the document]
Comment (amickan) [reply app/tests/challenges_tests/test_tasks.py:157]: This is fine, I think this is how we test all our periodic tasks.

File: app/grandchallenge/challenges/emails.py [hunk 5]
```diff
@@ -91,3 +91,24 @@ def send_challenge_status_update_email(challengerequest, challenge=None):
         recipients=[challengerequest.creator],
         subscription_type=EmailSubscriptionTypes.SYSTEM,
     )
+
+
+def send_email_percent_budget_consumed_alert(challenge, warning_threshold):
+    send_standard_email_batch(
+        site=Site.objects.get_current(),
+        subject=format_html(
+            "[{challenge_name}] Challenge {warning_threshold}% Budget Consumed Alert",
```
Comment (amickan) [inline app/grandchallenge/challenges/emails.py:100]: I think Challenge is redundant here, right? 

```suggestion
            "[{challenge_name}] {warning_threshold}% Budget Consumed Alert",
```

File: app/grandchallenge/challenges/tasks.py (hunk 3 above)
Comment (pkcakeout) [reply app/grandchallenge/challenges/tasks.py:75]: Just to confirm: I also find that hard to read. Black did not do this statement justice. All the operators "seem to be equal" even though this is an example of the rare ternary comparison... hard to parse!

File: app/grandchallenge/challenges/tasks.py [hunk 6]
```diff
@@ -55,12 +58,34 @@ def update_challenge_results_cache():
...
+
+
 @acks_late_2xlarge_task
 def update_compute_costs_and_storage_size():
     challenges = Challenge.objects.all()
 
-    for challenge in challenges:
+    for challenge in challenges.with_available_compute():
         annotate_compute_costs_and_storage_size(challenge=challenge)
+        send_alert_if_budget_consumed_warning_threshold_exceeded(
```
Comment (jmsmkn) [inline app/grandchallenge/challenges/tasks.py:86]: I think this task should be done in a single transaction as creating the email objects could cause the emails to be sent out when the challenges have not been updated, so everything needs to be updated at the same time. You can do this by adding `@transaction.atomic` on the line below the `@acks_late_2xlarge_task`

File: app/tests/challenges_tests/test_tasks.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (jmsmkn) [reply app/tests/challenges_tests/test_tasks.py:157]: This is the way to do it!

**Review by amickan — COMMENTED** [review]
LGTM. Just some small comments. 

I do have one general question: we're only taking into account compute costs here, not storage costs. I see that this is also what is checked in `percent_budget_consumed` and is what is displayed on the cost overview. Yet, with every new submission, we presumably also store a new Docker image, which affects the storage costs. Are the Docker storage costs usually overestimated on the invoice and hence hard to exceed?

**Comment by amickan** [comment]
Oh I see that James and I looked at this in parallel. Go with James' comments, obviously. ;-)

**Comment by jmsmkn** [comment]
> Oh I see that James and I looked at this in parallel. Go with James' comments, obviously. ;-)

Yes it seems like the review tab does not auto update like the comments one. Luckily (or rather unsurprisingly) we gave the same direction!

**Comment by jmsmkn** [comment]
> I do have one general question: we're only taking into account compute costs here, not storage costs. I see that this is also what is checked in `percent_budget_consumed` and is what is displayed on the cost overview. Yet, with every new submission, we presumably also store a new Docker image, which affects the storage costs. Are the Docker storage costs usually overestimated on the invoice and hence hard to exceed?

We should take into account storage but do not. We have the old problem that the two fields are separate (like how we used to assign budget/submissions per phase). I don't think there is a reason to consider them separately now. There is the problem that the storage costs are an ongoing rate, whereas the compute costs are fixed in time. We don't have per-month tracking of storage costs, I think to really make it work we would need that.

File: app/grandchallenge/challenges/tasks.py (hunk 3 above)
Comment (koopmant) [reply app/grandchallenge/challenges/tasks.py:75]: Yeah, sorry about this one. I was optimizing the code and it got a bit out of hand. And then, as Paul said, Black came in and made matters even worse... I'll clean it up!

File: app/grandchallenge/challenges/tasks.py (hunk 6 above)
Comment (koopmant) [reply app/grandchallenge/challenges/tasks.py:86]: I had not realized that, thanks!

File: app/tests/challenges_tests/test_tasks.py [hunk 7]
```diff
@@ -121,3 +131,42 @@ def test_challenge_request_budget_calculation(settings):
         + challenge_request.budget["Total phase 2"]
         + challenge_request.budget["Docker storage cost"]
     )
+
+
+@pytest.mark.django_db
+def test_challenge_budget_alert_email():
```
[the remark this answers is not in the document]
Comment (koopmant) [reply app/tests/challenges_tests/test_tasks.py]: Wow, thanks a lot for this comment. It made me uncover a bunch of mistakes in the code!

File: app/grandchallenge/challenges/tasks.py (hunk 3 above)
Comment (koopmant) [reply app/grandchallenge/challenges/tasks.py]: It should be a lot better now.

File: app/grandchallenge/challenges/emails.py [hunk 8]
```diff
@@ -91,3 +91,24 @@ def send_challenge_status_update_email(challengerequest, challenge=None):
...
+            warning_threshold=warning_threshold,
+        ),
+        markdown_message=format_html(
+            "We would like to inform you that {percent_budget_consumed}% of the "
+            "compute budget for your challenge has been used.",
+            percent_budget_consumed=challenge.percent_budget_consumed,
+        ),
+        recipients=[
+            *challenge.get_admins(),
+            *get_user_model().objects.filter(is_staff=True, is_active=True),
```
[the remark this answers is not in the document]
Comment (koopmant) [reply app/grandchallenge/challenges/emails.py]: Thank you, used it now.

File: app/grandchallenge/challenges/models.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (koopmant) [reply app/grandchallenge/challenges/models.py]: Good points. Reverted this.

File: app/tests/challenges_tests/test_tasks.py [hunk 9]
```diff
@@ -121,3 +131,94 @@ def test_challenge_request_budget_calculation(settings):
         + challenge_request.budget["Total phase 2"]
         + challenge_request.budget["Docker storage cost"]
     )
+
+
+@pytest.mark.django_db
+def test_challenge_budget_alert_email(settings):
```
[the remark this answers is not in the document]
Comment (koopmant) [reply app/tests/challenges_tests/test_tasks.py:137]: We do not. I will add a test for it.

File: app/grandchallenge/challenges/models.py [hunk 10]
```diff
@@ -376,6 +380,20 @@ class Challenge(ChallengeBase):
...
     )
 
+    percent_budget_consumed_warning_thresholds = models.JSONField(
+        default=get_default_percent_budget_consumed_warning_thresholds,
+        validators=[
+            JSONValidator(
+                schema={
+                    "$schema": "http://json-schema.org/draft-07/schema",
+                    "type": "array",
+                    "items": {"type": "integer"},
```
[the remark this answers is not in the document]
Comment (koopmant) [reply app/grandchallenge/challenges/models.py:390]: Perhaps a silly question, but would a limit of more than 100% ever be used? Because that would still 
work.

[the remark this answers is not in the document]
Comment (koopmant) [reply app/grandchallenge/challenges/models.py:390]: Btw, `exclusiveMinimum` is only a boolean in JSON draft 4. https://json-schema.org/understanding-json-schema/reference/numeric#range

**Review by jmsmkn — APPROVED** [review]

File: app/grandchallenge/challenges/models.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (jmsmkn) [reply app/grandchallenge/challenges/models.py]: Good to know!
